### PR TITLE
xds: implement XdsClient APIs for watching LDS/RDS resources individually

### PR DESCRIPTION
**Pull request by voidzcy**
First take for replacing ConfigWatcher with LdsResourceWatcher and RdsResourceWatcher. The next take is #7469.

Effective changes should be for LDS/RDS protocols only. Tests for XdsClientImpl are cleaned up.

### CODE REVIEW

**Review by dapengzhang0 — COMMENTED** [review]
Posted diff between `XdsClientImpl` and `XdsClientImpl2` at 1323354 for convenience.

File: xds/src/main/java/io/grpc/xds/XdsClient.java [hunk 1]
```diff
@@ -206,7 +211,8 @@ LdsUpdate build() {
     private final List<VirtualHost> virtualHosts;
 
     private RdsUpdate(List<VirtualHost> virtualHosts) {
-      this.virtualHosts = virtualHosts;
+      this.virtualHosts =
+          Collections.unmodifiableList(new ArrayList<>(checkNotNull(virtualHosts)));
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClient.java]: Fixed.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl2.java [hunk 2]
```diff
@@ -0,0 +1,1688 @@
...
+    logger.log(XdsLogLevel.INFO, "Started watching listener for port {0}", port);
+    if (rpcRetryTimer != null && rpcRetryTimer.isPending()) {
+      // Currently in retry backoff.
+      return;
+    }
+    if (adsStream == null) {
+      startRpcStream();
+    }
+    updateNodeMetadataForListenerRequest(port);
+    adsStream.sendXdsRequest(ResourceType.LDS, ImmutableList.<String>of());
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl2.java:314]: Seems this request is never retried for the `listenerWatcher` case.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl2.java [hunk 3]
```diff
@@ -0,0 +1,1688 @@
...
+        }
+      }
+    }
+
+    void onAbsent() {
+      if (respTimer != null && respTimer.isPending()) {  // too early to conclude absence
+        return;
+      }
+      logger.log(XdsLogLevel.INFO, "Conclude {0} resource {1} not exist", type, resource);
+      boolean oldAbsent = absent;
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl2.java:1099]: Same comment as https://github.com/grpc/grpc-java/pull/7458#discussion_r496167917

File: xds/src/main/java/io/grpc/xds/XdsClientImpl2.java [hunk 4]
```diff
@@ -0,0 +1,1688 @@
...
+
+    void onData(ResourceUpdate data) {
+      if (respTimer != null && respTimer.isPending()) {
+        respTimer.cancel();
+        respTimer = null;
+      }
+      ResourceUpdate oldData = this.data;
+      this.data = data;
+      absent = false;
+      if (!Objects.equals(oldData, data)) {
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl2.java:1087]: Still the https://github.com/grpc/grpc-java/pull/7458#discussion_r496203097 on `RdsUpdate.equals()`

File: xds/src/main/java/io/grpc/xds/XdsClientImpl2.java [hunk 5]
```diff
@@ -0,0 +1,1688 @@
...
+    private void handleStreamClosed(Status error) {
+      checkArgument(!error.isOk(), "unexpected OK status");
+      if (closed) {
+        return;
+      }
+      logger.log(
+          XdsLogLevel.ERROR,
+          "ADS stream closed with status {0}: {1}. Cause: {2}",
+          error.getCode(), error.getDescription(), error.getCause());
+      closed = true;
```
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/XdsClientImpl2.java:1313]: Why deleted
```java
if (listenerWatcher != null) {
  listenerWatcher.onError(error);
}
```

File: xds/src/main/java/io/grpc/xds/XdsClientImpl2.java (hunk 3 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl2.java]: Oops, sorry sorry. I was working on this PR in parallel with the previous one, forgot to fix the same thing in this one.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl2.java (hunk 4 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl2.java:1099]: Added.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl2.java (hunk 5 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl2.java:1324]: Oops, forgot Sanjay's code. Added back now.

File: xds/src/main/java/io/grpc/xds/XdsClientImpl2.java (hunk 2 above)
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/XdsClientImpl2.java:318]: Forgot that was for Sanjay's code. Added back.

**Review by dapengzhang0 — APPROVED** [review]

File: xds/src/test/java/io/grpc/xds/XdsClientImplTest2.java [hunk 6]
```diff
@@ -0,0 +1,1857 @@
...
+        @SuppressWarnings("unchecked")
+        StreamObserver<DiscoveryRequest> requestObserver = mock(StreamObserver.class);
+        RpcCall<DiscoveryRequest, DiscoveryResponse> call =
+            new RpcCall<>(requestObserver, responseObserver);
+        resourceDiscoveryCalls.offer(call);
+        Context.current().addListener(
+            new CancellationListener() {
+              @Override
+              public void cancelled(Context context) {
+                adsEnded.set(true);
```
[the remark this answers is not in the document]
Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/XdsClientImplTest2.java:238]: You are right. Added.

**Pull request merged**